File: uva10131/uva10131.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class e{
	public:
		int w,iq,idx;
};

bool cmp(const e& x,const e& y){
		return x.iq > y.iq;
}

vector<e> inp;

void solve(){
	int N = inp.size(),i,j,maxv = 1,maxi = 0;
	vector<vector<int> > path(N);
	path[0].push_back(inp[0].idx);
	for(i = 1; i < N; i++){
		for(j = 0; j < i;j++){
			if(inp[j].w < inp[i].w && inp[j].iq > inp[i].iq && path[i].size() < path[j].size()+1)
				path[i] = path[j];
		}
		path[i].push_back(inp[i].idx);
		if(path[i].size() > maxv){
			maxv = path[i].size();
			maxi = i;
		}

	}
	cout<<path[maxi].size()<<endl;
	for(i = 0; i < path[maxi].size();i++)
		cout<<path[maxi][i]<<endl;

}

int main(){
	int w,iq,i = 1;
	e tmp;
	while(cin>>w>>iq){
		tmp.w = w;
		tmp.iq = iq;
		tmp.idx = i;
		inp.push_back(tmp);
		i += 1;
	}
	sort(inp.begin(),inp.end(),cmp);
	solve();

	return 0;
}

File: uva10131/uva10131.py
from sys import stdin

def solve():
    N = len(W)
    path,maxv,maxi = [[] for _  in range(N)],1,0
    path[0].append(idx[0])
    for i in range(1,N):
        for j in range(0,i):
            if W[j] < W[i] and IQ[j] > IQ[i] and len(path[i]) < len(path[j])+1:
                path[i] = list(path[j])
        path[i].append(idx[i])
        if len(path[i]) > maxv:
            maxv = len(path[i])
            maxi = i
    return path[maxi]


def main():
    global W,IQ,idx
    inp,txt,W,IQ,L,idx = stdin.readlines(),list(),list(),list(),list(),list()
    for line in inp:
        if line.strip() != "\n" and line.strip() != "":
            txt.append(line.strip())
    i = 1
    for line in txt:
        w,iq = map(int,line.split())
        L.append((w,iq,i))
        i += 1
    L.sort(key = lambda x: -x[1])
    for w,iq,i in L:
        #print(w,iq,i)
        W.append(w)
        IQ.append(iq)
        idx.append(i)
    
    ans = solve()
    print(len(ans))
    for i in range(len(ans)):
        print(ans[i])
    
main()
